File: apps/windows/TeslaSync.App/dashboard-widgets/APIUsageWidget/APIUsageWidget.ViewModel.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.DashboardWidgets;

/// <summary>
/// The data port the <see cref="ApiUsageViewModel"/> binds to (P1/S8 state-holder seam). It yields the
/// cache-then-network sequence of parsed API call-log stats for <c>GET /api-logs/stats</c> — the native
/// analogue of the web <c>useApiLogStats</c> hook. The view never performs HTTP itself; the concrete
/// <see cref="ApiUsageSource"/> (or a test fake) drives this.
/// </summary>
public interface IApiUsageSource
{
    /// <summary>Stream the cache-then-network API-usage snapshots, newest cache first.</summary>
    IAsyncEnumerable<RepositoryResult<ApiUsageStats>> StreamAsync(CancellationToken cancellationToken = default);
}

/// <summary>
/// Canonical registry metadata for the API Usage surface — the native mirror of the web registry entry
/// in web/src/features/dashboard/widgets/registry/system.ts. The dashboard grid system binds this
/// surface with the same <see cref="Id"/> and honours the same size constraints.
/// </summary>
public static class ApiUsageRegistration
{
    /// <summary>Stable registry id (matches the web registry).</summary>
    public const string Id = "api-usage";

    /// <summary>Widget category (matches the web registry).</summary>
    public const string Category = "system";

    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "APIUsageWidget";

    /// <summary>Default footprint: 2 columns × 2 rows.</summary>
    public static ApiUsageSize DefaultSize => new(2, 2);

    /// <summary>Minimum footprint: 1 column × 2 rows.</summary>
    public static ApiUsageSize MinSize => new(1, 2);

    /// <summary>Maximum footprint: 4 columns × 40 rows.</summary>
    public static ApiUsageSize MaxSize => new(4, 40);

    /// <summary>Localized display name (web registry "API Usage").</summary>
    public static string Name(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer.GetString("widget.apiUsage.title", "API Usage");
    }

    /// <summary>Localized description (web registry copy).</summary>
    public static string Description(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer.GetString(
            "widget.apiUsage.description",
            "API call volume, response times, error rates, top endpoints");
    }

    /// <summary>True when <paramref name="size"/> falls within the min/max footprint constraints.</summary>
    public static bool IsWithinBounds(ApiUsageSize size) =>
        size.Cols >= MinSize.Cols && size.Cols <= MaxSize.Cols &&
        size.Rows >= MinSize.Rows && size.Rows <= MaxSize.Rows;

    /// <summary>Clamp <paramref name="size"/> into the supported min/max footprint.</summary>
    public static ApiUsageSize Clamp(ApiUsageSize size) => new(
        Math.Clamp(size.Cols, MinSize.Cols, MaxSize.Cols),
        Math.Clamp(size.Rows, MinSize.Rows, MaxSize.Rows));
}

/// <summary>
/// PII-safe diagnostics for the API Usage surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never an endpoint path, request count or
/// error message — so a diagnostics line can never leak operational data. Thread-safe.
/// </summary>
public sealed class ApiUsageDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public ApiUsageDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=APIUsageWidget</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={ApiUsageRegistration.Slug}");
    }
}

/// <summary>
/// UI-thread-free state holder backing the WinUI <see cref="APIUsageWidget"/> view — the native port of
/// the web <c>APIUsageWidget</c>'s hook composition
/// (web/src/features/dashboard/widgets/APIUsageWidget.tsx). It consumes the cache-then-network
/// <see cref="IApiUsageSource"/>, projects each snapshot through <see cref="ApiUsageProjection"/>, and
/// exposes the mutually-exclusive <see cref="State"/> plus the header freshness flags so the view is a
/// thin renderer. Drive it from one confinement (the UI thread); it is not internally synchronised.
/// </summary>
public sealed class ApiUsageViewModel : INotifyPropertyChanged, IDisposable
{
    private readonly IApiUsageSource _source;
    private readonly ILocalizer _localizer;
    private readonly Func<DateTimeOffset> _clock;

    private ApiUsageSize _size;
    private CancellationTokenSource? _cts;
    private RepositoryResult<ApiUsageStats>? _last;
    private bool _disposed;

    private ApiUsageState _state = ApiUsageState.Loading;
    private ApiUsageDisplay _display;
    private DateTimeOffset? _updatedAt;
    private bool _isFetching;
    private bool _isError;
    private bool _isStale;
    private string? _errorMessage;
    private int _attempts;

    /// <summary>Creates the holder over its data source, localizer and footprint.</summary>
    public ApiUsageViewModel(
        IApiUsageSource source,
        ILocalizer localizer,
        ApiUsageSize size,
        Func<DateTimeOffset>? clock = null)
    {
        ArgumentNullException.ThrowIfNull(source);
        ArgumentNullException.ThrowIfNull(localizer);
        _source = source;
        _localizer = localizer;
        _size = size;
        _clock = clock ?? (() => DateTimeOffset.Now);
        _display = ApiUsageProjection.Project(ApiUsageStats.Empty, _size, _localizer);
    }

    /// <inheritdoc />
    public event PropertyChangedEventHandler? PropertyChanged;

    /// <summary>The current mutually-exclusive surface state.</summary>
    public ApiUsageState State
    {
        get => _state;
        private set => Set(ref _state, value);
    }

    /// <summary>The projected, render-ready display model (stats + compact hero).</summary>
    public ApiUsageDisplay Display
    {
        get => _display;
        private set
        {
            _display = value;
            Raise(nameof(Display));
            Raise(nameof(HasData));
        }
    }

    /// <summary>Last successful update timestamp surfaced in the header freshness chip.</summary>
    public DateTimeOffset? UpdatedAt
    {
        get => _updatedAt;
        private set => Set(ref _updatedAt, value);
    }

    /// <summary>True while a background refresh is in flight (header chip pulses).</summary>
    public bool IsFetching
    {
        get => _isFetching;
        private set => Set(ref _isFetching, value);
    }

    /// <summary>True when the last load failed (drives the error surface + header chip).</summary>
    public bool IsError
    {
        get => _isError;
        private set => Set(ref _isError, value);
    }

    /// <summary>True when the shown snapshot is older than the freshness window.</summary>
    public bool IsStale
    {
        get => _isStale;
        private set => Set(ref _isStale, value);
    }

    /// <summary>Localized error message shown in the error surface.</summary>
    public string? ErrorMessage
    {
        get => _errorMessage;
        private set => Set(ref _errorMessage, value);
    }

    /// <summary>Number of load attempts started (including retries).</summary>
    public int Attempts
    {
        get => _attempts;
        private set => Set(ref _attempts, value);
    }

    /// <summary>True when the snapshot has a payload to render (web <c>data</c> truthiness).</summary>
    public bool HasData => _display.HasData;

    /// <summary>Localized widget title (web <c>widget.apiUsage.title</c>).</summary>
    public string Title => ApiUsageRegistration.Name(_localizer);

    /// <summary>Localized empty-state message (web <c>widget.apiUsage.noData</c>).</summary>
    public string EmptyMessage => _localizer.GetString("widget.apiUsage.noData", "No API usage data");

    /// <summary>The widget footprint; reassigning re-projects the current snapshot for the new layout.</summary>
    public ApiUsageSize Size
    {
        get => _size;
        set
        {
            if (_size == value)
            {
                return;
            }

            _size = value;
            Raise(nameof(Size));
            Reproject();
        }
    }

    /// <summary>
    /// Run a cache-then-network load: counts the attempt, shows the skeleton only when nothing is already
    /// visible (otherwise keeps content while refreshing), and folds every emission into
    /// <see cref="State"/> + <see cref="Display"/>. A superseding load cancels the prior one.
    /// </summary>
    public async Task LoadAsync(CancellationToken cancellationToken = default)
    {
        var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        var previous = Interlocked.Exchange(ref _cts, cts);
        previous?.Cancel();
        previous?.Dispose();

        Attempts++;
        if (!HasContent())
        {
            SetLoading();
        }
        else
        {
            IsFetching = true;
        }

        try
        {
            await foreach (var result in _source.StreamAsync(cts.Token).ConfigureAwait(false))
            {
                Apply(result);
            }
        }
        catch (OperationCanceledException)
        {
            // Superseded by a newer load (or disposed) — drop this emission silently.
        }
    }

    /// <summary>Retry after a failure — re-runs the load from the top.</summary>
    public Task RetryAsync() => LoadAsync();

    /// <inheritdoc />
    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        var cts = Interlocked.Exchange(ref _cts, null);
        cts?.Cancel();
        cts?.Dispose();
        GC.SuppressFinalize(this);
    }

    private bool HasContent() =>
        _state is ApiUsageState.Loaded or ApiUsageState.Stale or ApiUsageState.Offline;

    private void Apply(RepositoryResult<ApiUsageStats> result)
    {
        _last = result;
        switch (result.Status)
        {
            case LoadStatus.Loading:
                if (!HasContent())
                {
                    SetLoading();
                }

                IsFetching = true;
                break;

            case LoadStatus.Cached:
                ApplySnapshot(result.Value!, result.FetchedAt, result.IsStale, fetching: false, error: null);
                break;

            case LoadStatus.Refreshing:
                ApplySnapshot(result.Value!, result.FetchedAt, result.IsStale, fetching: true, error: null);
                break;

            case LoadStatus.Loaded:
                ApplySnapshot(result.Value!, result.FetchedAt, stale: false, fetching: false, error: null);
                break;

            case LoadStatus.Empty:
                SetEmpty(result.FetchedAt);
                break;

            case LoadStatus.Offline:
                ApplySnapshot(result.Value!, result.FetchedAt, stale: true, fetching: false, error: result.Error, offline: true);
                break;

            default:
                SetError(result.Error);
                break;
        }
    }

    private void ApplySnapshot(
        ApiUsageStats stats,
        DateTimeOffset? fetchedAt,
        bool stale,
        bool fetching,
        RepositoryError? error,
        bool offline = false)
    {
        var display = ApiUsageProjection.Project(stats, _size, _localizer);
        Display = display;

        if (!stats.HasData)
        {
            SetEmpty(fetchedAt, keepDisplay: true);
            return;
        }

        UpdatedAt = fetchedAt;
        IsFetching = fetching;
        IsStale = stale;
        IsError = false;
        ErrorMessage = offline ? ErrorTextFor(error) : null;
        State = offline ? ApiUsageState.Offline : stale ? ApiUsageState.Stale : ApiUsageState.Loaded;
    }

    private void Reproject()
    {
        if (_last is { HasValue: true } last)
        {
            Apply(last);
        }
        else
        {
            Display = ApiUsageProjection.Project(ApiUsageStats.Empty, _size, _localizer);
        }
    }

    private void SetLoading()
    {
        IsError = false;
        ErrorMessage = null;
        State = ApiUsageState.Loading;
    }

    private void SetEmpty(DateTimeOffset? fetchedAt, bool keepDisplay = false)
    {
        if (!keepDisplay)
        {
            Display = ApiUsageProjection.Project(ApiUsageStats.Empty, _size, _localizer);
        }

        UpdatedAt = fetchedAt;
        IsFetching = false;
        IsStale = false;
        IsError = false;
        ErrorMessage = null;
        State = ApiUsageState.Empty;
    }

    private void SetError(RepositoryError? error)
    {
        IsFetching = false;
        IsStale = false;
        IsError = true;
        ErrorMessage = ErrorTextFor(error);
        State = ApiUsageState.Error;
    }

    private string ErrorTextFor(RepositoryError? error)
    {
        string key = error?.Kind switch
        {
            RepositoryErrorKind.Unauthorized => "widget.apiUsage.error.auth",
            RepositoryErrorKind.Offline or RepositoryErrorKind.Network => "widget.apiUsage.error.offline",
            _ => "widget.apiUsage.error",
        };

        string fallback = error?.Kind switch
        {
            RepositoryErrorKind.Unauthorized => "Sign in to view API usage",
            RepositoryErrorKind.Offline or RepositoryErrorKind.Network => "You're offline — showing the last cached API usage",
            _ => "Couldn't load API usage",
        };

        return _localizer.GetString(key, fallback);
    }

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
        {
            return;
        }

        field = value;
        Raise(name);
    }

    private void Raise(string? name) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

File: apps/android/app/src/main/java/io/teslasync/android/dashboard/widgets/APIUsageWidget.ViewModel.kt
@file:Suppress("MatchingDeclarationName")

package io.teslasync.android.dashboard.widgets

import io.teslasync.android.data.BaseFeedViewModel
import io.teslasync.android.data.UiState
import io.teslasync.shared.core.data.repo.AdminRepository
import io.teslasync.shared.core.data.repo.Resource
import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.update

/**
 * The data port the [ApiUsageViewModel] binds to — the native analogue of the web `useApiLogStats`
 * hook. It yields the cache-then-network sequence of parsed API call-log stats for
 * `GET /api-logs/stats`. The view never performs HTTP itself; the production [AdminApiUsageSource]
 * (or a test double) drives this seam.
 */
fun interface ApiUsageSource {
    /** Stream the cache-then-network API-usage snapshots, cached value first. */
    fun stream(): Flow<Resource<ApiUsageStats>>
}

/**
 * The repository-backed [ApiUsageSource]. It runs one cache-then-network read of
 * `GET /api-logs/stats` through the shared [AdminRepository] (the cross-platform `useAdmin` data
 * port) and parses each [Resource] emission into an [ApiUsageStats] via [ApiUsageStats.fromJson],
 * preserving every freshness flag (cached / refreshing / stale / offline / error) so the view-model
 * can render the full state matrix. Re-collecting the cold repository flow on retry performs a fresh
 * cache-then-network read, so the error/offline retry affordance genuinely re-fetches.
 */
class AdminApiUsageSource(
    private val repository: AdminRepository,
) : ApiUsageSource {
    override fun stream(): Flow<Resource<ApiUsageStats>> =
        repository.apiLogStats().map { resource -> resource.mapValue(ApiUsageStats::fromJson) }
}

/**
 * Maps a [Resource]'s payload (when present) while preserving its status and freshness stamps. Kept
 * pure so the parse-and-preserve contract is unit-tested without a network or cache.
 */
fun <T, R> Resource<T>.mapValue(transform: (T) -> R): Resource<R> =
    when (this) {
        is Resource.Loading -> Resource.Loading(cached?.let(transform), fetchedAt, stale)
        is Resource.Success -> Resource.Success(transform(data), fetchedAt, stale)
        is Resource.Error -> Resource.Error(cached?.let(transform), fetchedAt, stale, error)
    }

/**
 * PII-safe diagnostics for the API Usage surface (P1/S11 diagnostics contract). Records only the
 * operational `view.opened` event with the surface slug through the single sanctioned redacting
 * [Logger] — never an endpoint path, request count, or error message — so a diagnostics line can
 * never leak operational data.
 */
class ApiUsageDiagnostics(
    private val logger: Logger,
) {
    /** Record that the surface was opened, emitting `view.opened slug=APIUsageWidget`. */
    fun recordViewOpened() {
        logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SLUG to ApiUsageRegistration.SLUG))
    }

    private companion object {
        const val EVENT_VIEW_OPENED = "view.opened"
        const val FIELD_SLUG = "slug"
    }
}

/**
 * Lifecycle-aware state holder backing the [APIUsageWidget] surface — the native port of the web
 * `APIUsageWidget`'s hook composition. It consumes the cache-then-network [ApiUsageSource] and
 * exposes the projected [UiState] so the Composable is a thin renderer; the view performs no HTTP.
 *
 * [refresh] re-runs the read by bumping a trigger that `flatMapLatest`-re-subscribes to a fresh
 * [ApiUsageSource.stream] — backing the pull-to-refresh and the error/offline retry affordance.
 *
 * @param source the cache-then-network data port (production: [AdminApiUsageSource]).
 * @param logger the single sanctioned redacting logger (ADR-016).
 * @param scope test seam; production passes nothing and uses `viewModelScope`.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class ApiUsageViewModel(
    private val source: ApiUsageSource,
    logger: Logger,
    scope: CoroutineScope? = null,
) : BaseFeedViewModel(logger, scope) {
    private val refreshSignal = MutableStateFlow(0)

    /** The API-usage rollup as cache-then-network UI state (loading / content / empty / stale / error). */
    val stats: StateFlow<UiState<ApiUsageStats>> =
        refreshSignal
            .flatMapLatest { source.stream() }
            .asUiState(isEmpty = { !it.hasData })

    /** Re-fetches the rollup (pull-to-refresh / retry). A no-op while nobody observes [stats]. */
    fun refresh() {
        logger.info(EVENT_REFRESH)
        refreshSignal.update { it + 1 }
    }

    private companion object {
        const val EVENT_REFRESH = "api_usage.refresh"
    }
}
